### ISSUE TITLE
How to create meta.json for Liboro dataset?

### ISSUE THREAD

**Issue by choucaicai** [2toinf/X-VLA#66]

**Comment by Saberlve** [comment]
You can check the dataset handler code to understand what would be used in the meta.json. And construct the keys.

**Comment by CathyF9600** [comment]
Hello, here is an example meta file I'm using. You can adjust the field based on keys in your hdf5 dataset.
```
{
  "dataset_name": "libero",
  "language_instruction_key": "language_instruction",
  "observation_key": [
    "observation/third_image",
    "observation/wrist_image"
  ],
  "datalist":[
    "path/demo1.hdf5",
    ...
  ]
}
```

**Comment by choucaicai** [comment]
Hello, how is your action representation concatenated into 10 dimensions? Do you need to convert ee_ori to 6D? I see in the code that the action needs to be 10-dimensional: ee pos + ee ori 6d + gripper.  action. I use the original LIBERO datasets and split each task's hdf5 file into demo1.hdf5, demo2.hdf5, etc. I extract ee_pos and ee_ori as my action representation. However, I found that during evaluation, the success rate is 0. I suspect there might be an issue with my data conversion.

Full convrt Code:
```
import h5py
import os, json
from pathlib import Path
import numpy as np
import h5py
import os, json
from pathlib import Path
import numpy as np
import h5py
import re
import tqdm
import  cv2

def axis_angle_to_rotmat_batch(aa: np.ndarray) -> np.ndarray:
    """
    aa: [T,3] axis-angle rotation vector (axis * theta), theta in radians
    return: [T,3,3] rotation matrices
    """
    aa = aa.astype(np.float64)
    T = aa.shape[0]
    theta = np.linalg.norm(aa, axis=1, keepdims=True)  # [T,1]
    # avoid div0
    axis = aa / (theta + 1e-12)                        # [T,3]

    ax, ay, az = axis[:, 0], axis[:, 1], axis[:, 2]
    zeros = np.zeros(T, dtype=np.float64)

    K = np.stack([
        zeros, -az,   ay,
        az,    zeros, -ax,
        -ay,   ax,    zeros
    ], axis=1).reshape(T, 3, 3)

    I = np.eye(3, dtype=np.float64)[None, :, :]       # [1,3,3]
    sin = np.sin(theta)[:, None]                      # [T,1,1]
    cos = np.cos(theta)[:, None]                      # [T,1,1]

    # Rodrigues: R = I + sinθ K + (1-cosθ) K^2
    K2 = K @ K
    R = I + sin * K + (1.0 - cos) * K2

    # small-angle fallback: when theta ~ 0, use I
    small = (theta[:, 0] < 1e-8)
    if np.any(small):
        R[small] = np.eye(3, dtype=np.float64)

    return R


def rotmat_to_rot6d_batch(R: np.ndarray) -> np.ndarray:
    """
    R: [T,3,3]
    return rot6d: [T,6] = concat(R[:, :, 0], R[:, :, 1])
    matches your Mat_to_Rotate6D (take first two columns)
    """
    return np.concatenate([R[:, :, 0], R[:, :, 1]], axis=1)


def parse_language_from_filename(stem: str) -> str:
    # 1) remove trailing _demo if present
    s = re.sub(r"_demo$", "", stem)

    # 2) strip "<prefix>_SCENE<digits>_" and keep the instruction part
    m = re.match(r"^.*?_SCENE\d+_(.+)$", s)
    if m:
        return m.group(1)
    # fallback: return as-is (without _demo)
    return s


def gripper_states_to_grip_raw(gripper_states: np.ndarray) -> np.ndarray:
    """
    gripper_states: [T,2] (often two finger positions)
    We need grip_raw [T,1] such that >0 means CLOSED (to match your LiberoHandler).
    Heuristic: smaller finger opening => closed.
    """
    g = np.asarray(gripper_states)
    if g.ndim == 1:
        g = g[:, None]
    opening = g.mean(axis=1)  # [T]
    # threshold needs to be adjusted if your gripper scale differs
    closed = (opening < 0.02)
    grip_raw = np.where(closed, 1.0, -1.0).astype(np.float32)[:, None]  # [T,1]
    return grip_raw

def _flip_agentview(img: np.ndarray) -> np.ndarray:
    """Match original code behavior: vertical+horizontal flips."""
    return np.flip(np.flip(img, 0), 1)

def convert_dir(src_dir: str, out_dir: str, meta_json_path: str, dataset_name="LIBERO_10"):
    src_dir = Path(src_dir)
    out_dir = Path(out_dir)
    data_dir = out_dir / 'data'
    out_dir.mkdir(parents=True, exist_ok=True)
    data_dir.mkdir(parents=True, exist_ok=True)

    datalist = []
    ep_count = 0
    hd_files = sorted(src_dir.glob("*.hdf5"))
    
    for idx,src_path in enumerate(hd_files):
        print(f'Start {idx} / {len(hd_files)}')
        stem = src_path.stem
        lang_default = parse_language_from_filename(stem)
        print('lang_default:',lang_default)
        with h5py.File(src_path, "r") as f:
            demos = sorted(list(f["data"].keys()))
            for demo in tqdm.tqdm(demos):
                
                grp = f[f"data/{demo}"]
                obs = grp["obs"]

                agent = obs["agentview_rgb"][()]       # [T,128,128,3] uint8
                eih   = obs["eye_in_hand_rgb"][()]     # [T,128,128,3] uint8

                imgs_fliped = [_flip_agentview(img) for img in agent]
                agent = np.stack(imgs_fliped,axis=0)
                print('agent:',agent.shape)
                # for i_i,img in enumerate(agent):
                    # img224 = cv2.resize(img, (224, 224), interpolation=cv2.INTER_AREA)
                    # img224 = _flip_agentview(img224)
                    # cv2.imwrite(f'data_gen/{i_i:05d}.png',img224)
                    # cv2.imwrite(f'data_gen/0_bgr.png',img224_bgr)
                    
                ee_pos = obs["ee_pos"][()].astype(np.float32)   # [T,3]
                ee_ori = obs["ee_ori"][()]                      # [T,3] float64 -> axis-angle assumed
                ee_ori = ee_ori.astype(np.float64)
                # axis-angle -> rot6d
                R = axis_angle_to_rotmat_batch(ee_ori)          # [T,3,3]
                rot6d = rotmat_to_rot6d_batch(R).astype(np.float32)  # [T,6]
                # gripper -> grip_raw (positive means closed)
                grip_raw = gripper_states_to_grip_raw(obs["gripper_states"][()])  # [T,1]

                abs_action_6d = np.concatenate([ee_pos, rot6d, grip_raw], axis=1) # [T,10]
                assert abs_action_6d.shape[1] == 10
                # print('ee_pos:',ee_pos.shape,agent.shape)

                # write one-episode HDF5 (keys at ROOT, to match your LiberoHandler)
                out_path = data_dir / f"episode_{ep_count:06d}.hdf5"

                with h5py.File(out_path, "w") as fo:
                    fo.create_dataset("abs_action_6d", data=abs_action_6d,
                                      compression="gzip", compression_opts=4)
                    fo.create_dataset("language_instruction", data=np.string_(lang_default))
                    # keep full T; your LiberoHandler will drop first frame for images internally
                    fo.create_dataset("agentview_rgb", data=agent,
                                      compression="gzip", compression_opts=4)
                    fo.create_dataset("eye_in_hand_rgb", data=eih,
                                      compression="gzip", compression_opts=4)

                datalist.append(str(out_path.resolve()))
                ep_count += 1
        # break
    
    meta = {
        "dataset_name": dataset_name,
        # IMPORTANT: ensure registry can find LiberoHandler via robot_type
        "robot_type": "libero",
        "observation_key": ["agentview_rgb", "eye_in_hand_rgb"],
        "language_instruction_key": "language_instruction",
        "datalist": datalist
    }

    Path(meta_json_path).parent.mkdir(parents=True, exist_ok=True)
    with open(meta_json_path, "w") as f:
        json.dump(meta, f, indent=2, ensure_ascii=False)

    print(f"Done. episodes={ep_count}, meta={meta_json_path}")





if __name__ == "__main__":
    # libero_spatial
    convert_dir(
        src_dir="data/LIBERO-datasets/libero_spatial",
        out_dir="data/modified_libero_xvla/libero_spatial",
        meta_json_path="data/modified_libero_xvla/libero_spatial/meta.json",
        dataset_name="libero_spatial",
    )
```

**Comment by CathyF9600** [comment]
To generate 10-dim absolute EEF action, please see https://github.com/2toinf/X-VLA/blob/main/evaluation/libero/preprocess.md.

**Comment by jingjiajian** [comment]
When I fine-tuned on the X-VLA-Pt model using my own data, why is the evaluated trajectory basically the same every time? Even though the position of the object is different each time.

**Comment by CathyF9600** [comment]
I recommend checking your dataloader to see if it loads all of your training data. Furthermore, if you're using delta control you might have forgotten to denormalize during deployment.
